File: Angles/Angles.c
#include "Angles.h"


void setup_Angles()
{
    Angles_Mapping = OpenFileMapping(
        FILE_MAP_ALL_ACCESS,   // read/write access
        FALSE,                 // do not inherit the name
        Angles_Name);               // name of mapping object
    
    if (Angles_Mapping == NULL) {
        Angles_isMapped = false;
        return;
    }

    Angles_Buf = (LPTSTR)MapViewOfFile(Angles_Mapping, // handle to map object
        FILE_MAP_ALL_ACCESS,  // read/write permission
        0,
        0,
        BUF_SIZE);
    
    if (Angles_Buf == NULL) {
        Angles_isMapped = false;
        return;
    }
    Angles_isMapped = true;
}


float get_pitch()
{
    if (Angles_isMapped) {
        return Angles_struct.pitch;
    } else {
        return 0;
    }
}

float get_true_heading()
{
    if (Angles_isMapped) {
        return Angles_struct.true_heading;
    } else {
        return 0;
    }
}

float get_roll()
{
    if (Angles_isMapped) {
        return Angles_struct.roll;
    } else {
        return 0;
    }
}

void update_Angles() {
    if (Angles_isMapped) {
        memcpy(&Angles_struct, Angles_Buf, sizeof(Angles_struct));
    }
}

void close_Angles()
{
    if (Angles_isMapped) {
        
        UnmapViewOfFile(Angles_Buf);
        CloseHandle(Angles_Mapping);
    }
}

File: Angles/Angles.m
classdef Angles < matlab.System & coder.ExternalDependency
        % Angles- this block sends its inputs to active X-Plane simulation using shared memory. Note: in order for this block to work, special X-Plane plugin needs to be installed
    
        properties(Access = public)
        
        end
            
        properties(DiscreteState)
    
        end
    
        properties(Access = private)
    
        end
    
        methods(Access = protected)
            function setupImpl(obj)
                    coder.cinclude('Angles.h');
                    coder.ceval('setup_Angles');
            end
    
            function [pitch,true_heading,roll] = stepImpl(obj)
                pitch = single(0);
                true_heading = single(0);
                roll = single(0);
                coder.cinclude('Angles.h');
                coder.ceval('update_Angles');
                pitch = coder.ceval('get_pitch');
                true_heading = coder.ceval('get_true_heading');
                roll = coder.ceval('get_roll');
            end
    
            function resetImpl(obj)
                
            end
            
            function releaseImpl(obj)
                coder.cinclude('Angles.h');
                coder.ceval('close_Angles');
            end
        end
            
        methods (Static)
            function name = getDescriptiveName()
                name = 'Angles';
            end
            function b = isSupportedContext(context)
                b = context.isCodeGenTarget('rtw');
            end
            function updateBuildInfo(buildInfo, context)
            % Update the build-time buildInfo
                blockRoot = fileparts(mfilename('fullpath'));
                buildInfo.addIncludePaths({blockRoot});
                buildInfo.addIncludeFiles({'Angles.h'});
                buildInfo.addSourcePaths({blockRoot});
                buildInfo.addSourceFiles({'Angles.c'});
            end
        end
    end